src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,300 @@
+            let crop_window       = crop_window.translate_y(-height * 0.5);
+
+            let port_area_cropped = crop_window.intersection(port_area_aligned);
+
+            let color             = Var::<color::Rgba>::from("srgba(0.25,0.58,0.91,input_opacity)");
+            let port_area_colored = port_area_cropped.fill(color);
+
+            (port_area_colored + hover_area).into()
+        }
+    }
+}
+
+// ===========
+// === Frp ===
+// ===========
+
+
+type PortId = usize;
+
+#[derive(Clone,CloneRef,Debug)]
+pub struct Frp {
+    pub set_size        : frp::Source<Option<Vector2<f32>>>,
+    pub port_mouse_down : frp::Stream<PortId>,
+
+    on_port_mouse_down  : frp::Source<PortId>,
+}
+
+impl Frp {
+    fn new(network: &frp::Network) -> Self {
+        frp::extend! { network
+            def set_size           = source();
+            def on_port_mouse_down = source();
+
+            let port_mouse_down = on_port_mouse_down.clone_ref().into();
+        }
+        Self{set_size,on_port_mouse_down,port_mouse_down}
+    }
+}
+
+
+
+// =======================
+// === OutPutPortsData ===
+// =======================
+
+#[derive(Debug)]
+pub struct OutPutPortsData {
+    display_object : display::object::Instance,
+    scene          : Scene,
+    logger         : Logger,
+    size           : Vector2<f32>,
+    gap_width      : f32,
+    ports          : Vec<component::ShapeView<port_area::Shape>>,
+}
+
+impl OutPutPortsData {
+    fn init(mut self) -> Self {
+        self.update_shapes();
+        self
+    }
+
+    fn update_ports(&mut self) {
+        let port_num      = self.ports.len() as f32;
+
+        let width         = self.size.x;
+        let height        = self.size.y;
+        let element_width = width / port_num;
+        let element_size  = Vector2::new(element_width,height);
+
+        // Align shapes along width.
+        let x_start = -width/2.0 + NODE_SHAPE_PADDING;
+        let x_delta = element_width;
+        for (index, view) in self.ports.iter().enumerate(){
+            view.display_object().set_parent(&self.display_object);
+
+            let pos_x = x_start + x_delta * index as f32;
+            let pos_y = 0.0;
+            let pos   = Vector2::new(pos_x,pos_y);
+            view.set_position_xy(pos);
+
+            let shape = &view.shape;
+            shape.sprite.size().set(element_size);
+            shape.shape_width.set(width);
+            shape.padding.set(self.gap_width);
+            shape.offset_x.set(x_delta * index as f32);
+        }
+    }
+
+    fn update_shapes(&mut self) {
+        self.update_ports();
+    }
+
+    fn set_size(&mut self, size:Vector2<f32>) {
+        self.size = size;
+        self.update_shapes();
+    }
+}
+
+// ===================
+// === OutPutPorts ===
+// ===================
+
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    pub frp            : Frp,
+        network        : frp::Network,
+        data           : Rc<RefCell<OutPutPortsData>>,
```

Comment:
instead of putting so many variables in RefCell, but only the needed ones. A lot of borrow ping pong below is needed because of this design. In `OutPutPortsData` you have a lot of fields which are already internally mutable, like `scene`, `logger`, or `display_object`. Putting them in another `RefCell` is bad design. Moreover, why do you have multiple `display_object` for the same thing? Again because of this design. Use `RefCells` on the smallest parts possible.